uarray/_uarray_dispatch.cxx

```diff
@@ -179,14 +181,269 @@ std::string backend_to_domain_string(PyObject * backend) {
+            py_ref::ref(py_registered), BackendState::convert_backend);
+
+    return output;
+  }
+
+  static global_state_t convert_global_state(py_ref input) {
+    return convert_dict<
+        std::string, global_backends, decltype(BackendState::convert_domain),
+        decltype(BackendState::convert_global_backends)>(
+        input, BackendState::convert_domain,
+        BackendState::convert_global_backends);
+  }
+
+  static local_state_t convert_local_state(py_ref input) {
+    return convert_dict<
+        std::string, local_backends, decltype(BackendState::convert_domain),
+        decltype(BackendState::convert_local_backends)>(
+        input, BackendState::convert_domain,
+        BackendState::convert_local_backends);
+  }
+
+  static py_ref convert_py(py_ref input) { return input; }
+
+  static py_ref convert_py(backend_options input) {
+    py_ref output = py_ref::steal(Py_BuildValue(
+        "(ONN)", input.backend.get(), PyBool_FromLong(input.coerce),
+        PyBool_FromLong(input.only)));
+    if (!output)
+      throw std::runtime_error("");
+    return output;
+  }
+
+  static py_ref convert_py(const std::string & input) {
+    py_ref output =
+        py_ref::steal(PyUnicode_FromStringAndSize(input.c_str(), input.size()));
+    if (!output)
+      throw std::runtime_error("");
+    return output;
+  }
+
+  template <typename T>
+  static py_ref convert_py(const std::vector<T> & input) {
+    py_ref output = py_ref::steal(PyList_New(input.size()));
+
+    if (!output)
+      throw std::runtime_error("");
+
+    for (size_t i = 0; i < input.size(); i++) {
+      py_ref element = convert_py(input[i]);
+      Py_INCREF(element.get());
+      PyList_SET_ITEM(output.get(), i, element.get());
+    }
+
+    return output;
+  }
+
+  static py_ref convert_py(const local_backends & input) {
+    py_ref py_skipped = BackendState::convert_py(input.skipped);
+    py_ref py_preferred = BackendState::convert_py(input.preferred);
+    py_ref output = py_ref::steal(
+        Py_BuildValue("(OO)", py_skipped.get(), py_preferred.get()));
+
+    if (!output)
+      throw std::runtime_error("");
+
+    return output;
+  }
+
+  static py_ref convert_py(const global_backends & input) {
+    py_ref py_globals = BackendState::convert_py(input.global);
+    py_ref py_registered = BackendState::convert_py(input.registered);
+    py_ref output = py_ref::steal(
+        Py_BuildValue("(OO)", py_globals.get(), py_registered.get()));
+
+    if (!output)
+      throw std::runtime_error("");
+
+    return output;
+  }
+
+  template <typename K, typename V>
+  static py_ref convert_py(const std::unordered_map<K, V> & input) {
+    py_ref output = py_ref::steal(PyDict_New());
+
+    if (!output)
+      throw std::runtime_error("");
+
+    for (const auto & kv : input) {
+      py_ref py_key = convert_py(kv.first);
+      py_ref py_value = convert_py(kv.second);
+
+      Py_INCREF(py_key.get());
+      Py_INCREF(py_value.get());
```

Comment:
[`PyDict_SetItem`](https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItem) doesn't steal a reference so I don't think you should `Py_INCREF`.